### ISSUE TITLE
Support nameof keyword

### ISSUE THREAD

**Issue by maca88** [maca88/AsyncGenerator#52]
The following example:
```
public class TestCase
{
  [DefaultValue(nameof(Write))]
  public void Write()
  {
    SimpleFile.Write(nameof(DoSomething));
  }

  private void DoSomething()
  {
  }
}
```
when transfromed into a new type should generate:
```
public class TestCaseAsync
{
  [DefaultValue(nameof(WriteAsync))]
  public Task WriteAsync()
  {
    return SimpleFile.WriteAsync(nameof(DoSomething));
  }

  private void DoSomething()
  {
  }
}
```
where `nameof(Write)` should be converted to `nameof(WriteAsync)` and `DoSomething` must be copied as it is referenced by the argument of `WriteAsync` method

**Comment by fredericDelaporte** [comment]
> where `nameof(Write)` should be converted to `nameof(WriteAsync)`

Is there a test for this? It does not look to work, at least not with v0.8.1 and `nameof` used in function body, as in:

```c#
public class TestCase
{
  public void Write()
  {
    SimpleFile.Write(nameof(Write));
  }
}
```

When converted as new type, it generates:

```c#
public class TestCaseAsync
{
  public Task WriteAsync()
  {
    return SimpleFile.WriteAsync(nameof(Write));
  }
}
```

Which then fail compiling. (Actual case is [here](https://github.com/nhibernate/nhibernate-core/pull/1476#discussion_r155667971).)

**Comment by maca88** [comment]
> Is there a test for this?

For now there are only [two tests](https://github.com/maca88/AsyncGenerator/blob/master/Source/AsyncGenerator.Tests/Nameof/Fixture.cs), where one of them is related to a new type.

The generator currently produces a valid code for your `Write` example, the only problem is that an unneeded `try/catch` block is added, which will be addressed by [this issue](https://github.com/maca88/AsyncGenerator/issues/80).
The reason why you got an invalid code generated, is because the `nameof` argument is currently not converted, when inside and anonymous method passed as an argument, I opened [an issue](https://github.com/maca88/AsyncGenerator/issues/81) for this as well.

**Comment by fredericDelaporte** [comment]
Thanks, but it seems there is something else. The case I have linked does not occur inside an anonymous method, but in a variable initialization at the top of the method. In fact I have come back with this trouble because I had it again somewhere [else](https://github.com/nhibernate/nhibernate-core/pull/1547/files#diff-54b54d18d8f086fb5e6d3bcc9289c716R45). At this place `"Simple LINQ"` was originally `nameof("SimpleLinqPerf")`, and it was not converted too. (Tested again right now.)

And I have checked again the original case, changing this [line](https://github.com/nhibernate/nhibernate-core/pull/1476/files#diff-f5aae63b878973314c5c483b8b64c2eeR27) to `var region = nameof(RetrievedQueryCacheMatchesGloballyStoredOne);` does trigger the trouble too. So in both cases, `nameof` is not inside a lambda.

**Comment by maca88** [comment]
You are right, none of the two issues above is the one that your are hitting. It turns out that the actual problem lies in the `ignoreSearchForMethodReferences` setting that was [added for the test project](https://github.com/nhibernate/nhibernate-core/blob/master/src/AsyncGenerator.yml#L209). While the setting reduces the generation time for the test project (by 35 seconds on my comupter), it will cause a compilation error if one of the test/theory methods are used whitin a nameof keyword ([issue](https://github.com/maca88/AsyncGenerator/issues/82)).

[I've added](https://github.com/maca88/AsyncGenerator/commit/924d1db0e4f4d471359db6576fb7f2666c898f0b) an additional logic that will override the setting when one of those methods are found inside a `nameof` keyword. The fixes will be included in the version `0.8.3` that will be released shortly (waiting for Travis-CI to complete).

**Comment by fredericDelaporte** [comment]
Ok, thanks again!
